Oracle mode now uses DECIMAL with NEXTVAL

royvanrijn opened the pull request:
In our project we're using h2 as test substitute for our Oracle database.

When calling NEXTVAL in a query we ran into a compatibility problem between Oracle and h2 (even in 'Oracle mode'). H2 keeps returning Long/BigInteger while Oracle returns BigDecimal values.

In this commit I've tried to 'fix' the Oracle compatibility so the NEXTVAL function and SequenceValue is different in case of MODE=Oracle.

I wasn't able to succesfully run all the internal H2 tests, hopefully someone can verify this? We'd love to see this feature in a future H2 release.

Commit:
h2 and Oracle mode returns DECIMAL with SEQUENCES

When the mode is set to Oracle it now returns BigDecimal instead of
Long-values.

katzyn commented on h2/src/main/org/h2/expression/SequenceValue.java:
```diff
@@ -12,29 +12,46 @@
 public class SequenceValue extends Expression {
 
     private final Sequence sequence;
+    private final TypeInfo type;
```
This field is not needed. `sequence.getDatabase().getMode().decimalSequences` can be used everywhere.

katzyn reviewed (commented):
Thank you for your contribution!

Please, revert all unrelated changes in import statements. Better with force-pushing the branch to avoid all this noise in the history.

You need to add some tests.

You also need to update `h2/src/docsrc/html/features.html` with information about this deviation from Regular mode.

Please note that H2 may return either `java.sql.Types.DECIMAL` or `java.sql.Types.NUMERIC` depending on system settings. Most likely the `NUMERIC` will be used by default in the future.

Commit:
Using boolean instead of storing type

Changing the code for DECMIAL sequences and added documentation

Stupid IDE, stupid formatting

Undoing more formatting

Added test for NEXTVAL in Oracle mode

Commit:
Using boolean instead of storing type

Commit:
Changing the code for DECMIAL sequences and added documentation

grandinj commented:
> Can someone help me fix this and write the appropriate tests (as I'm unable to run them for some reason :-().

Have you followed our build instructions here:
   http://h2database.com/html/build.html

Possibly you are using the maven build, which doesn't work nearly as well.

and writing a test should be fairly easy once that works, just needs some extra code in TestCompatibilityOracle

Commit:
Stupid IDE, stupid formatting

grandinj commented on h2/src/main/org/h2/table/Column.java:
```diff
@@ -460,6 +460,8 @@ public void convertAutoIncrementToSequence(Session session, Schema schema,
         session.getDatabase().addSchemaObject(session, seq);
         setAutoIncrementOptions(null);
+
+        Mode mode = session.getDatabase().getMode();
```
leftover

katzyn replied:
This file should be completely reverted to its original state.

katzyn commented on h2/src/main/org/h2/engine/Mode.java:
```diff
@@ -5,14 +5,15 @@
+import org.h2.util.StringUtils;
+import org.h2.value.DataType;
+import org.h2.value.Value;
+
```
Move them into their original location.

katzyn commented on h2/src/main/org/h2/mode/FunctionsOracle.java:
```diff
@@ -17,6 +15,8 @@
 import org.h2.value.ValueUuid;
 
+import java.util.HashMap;
+
```
This import too.

Commit:
Undoing more formatting

katzyn commented:
@grandinj I think we will need to “Squash and merge” these changes instead of normal merge, because otherwise the `git blame Parser.java` will be completely messed up.

@royvanrijn
In `TestCompatibilityOracle` you can check the result of `ResultSet.getObject()` for simplicity. You can also check the metadata, but in that case you need to use `SysProperties.BIG_DECIMAL_IS_DECIMAL` to determine the expected data type (`DECIMAL` or `NUMERIC`).

grandinj commented:
@katzyn agreed

grandinj commented:
Nice work, thanks for addressing all the review comments

katzyn commented on h2/src/test/org/h2/test/db/TestCompatibilityOracle.java:
```diff
@@ -284,6 +286,25 @@ private void testDate() throws SQLException {
+        // Test NEXTVAL without Oracle MODE should return BIGINT
+        checkSequenceTypeWithMode("REGULAR", Types.BIGINT);
+        // Test NEXTVAL with Oracle MODE should return DECIMAL
+        checkSequenceTypeWithMode("Oracle", Types.DECIMAL);
```
The Oracle's JDBC driver returns `Types.NUMERIC`, isn't it?

As I already said, H2 returns either `Types.NUMERIC` or `Types.DECIMAL` depending on settings. I think that
```Java
SysProperties.BIG_DECIMAL_IS_DECIMAL ? Types.DECIMAL : Types.NUMERIC
```
should be used here to avoid failure when test suit is launched with non-default settings.

katzyn commented:
Please, also send a license statement as described here
https://h2database.com/html/build.html#providing_patches
to our mailing list
https://groups.google.com/forum/#!forum/h2-database

Commit:
Improved check for non-default settings (as requested)

Commit:
Merge remote-tracking branch 'h2database/master'

katzyn commented:
Thanks!

royvanrijn commented:
I’ve noticed the release-cycle of H2 is a bit... unpredictable haha. Sometimes weeks, right now 6 months.

Is there anything you can say about when this feature will be available in a release?

Right now we’re using a patched build ourself and I rather not have that 🙈

katzyn commented:
1.4.198 has many regressions, that's why 1.4.199 was released soon after it.
 
The next release is not planned yet and I have no idea when it will be built and published.

knutwannheden commented:
Just out of curiosity (and without looking at the patch in detail), doesn't maybe `CURRVAL` also need to be changed in the same way?
